engine/runtime/src/main/java/org/enso/interpreter/runtime/TruffleCompilerContext.java

```diff
@@ -181,11 +185,74 @@ public <T> Optional<T> loadCache(Cache<T, ?> cache) {
 
   @Override
   public <T> Optional<TruffleFile> saveCache(
-      Cache<T, ?> cache, T entry, boolean useGlobalCacheLocations) {
+          Cache<T, ?> cache, T entry, boolean useGlobalCacheLocations) {
     return cache.save(entry, context, useGlobalCacheLocations);
   }
 
+  @Override
+  public boolean typeContainsValues(String name) {
+    var type = context.getBuiltins().getBuiltinType(name);
+    return type != null && type.containsValues();
+  }
+
+  /**
+   * Lazy-initializes the IR for the builtins module.
+   */
+  @Override
+  public void initializeBuiltinsIr(
+          boolean irCachingEnabled, SerializationManager serializationManager,
+          FreshNameSupply freshNameSupply, Passes passes
+  ) {
+    var builtins = context.getBuiltins();
+    var builtinsModule = builtins.getModule();
+    if (!builtins.isIrInitialized()) {
+      log(
+              Level.FINE,
+              "Initialising IR for [{0}].",
+              builtinsModule.getName()
+      );
+
+      builtins.initializeBuiltinsSource();
+
+      if (irCachingEnabled) {
+        switch (serializationManager.deserialize(builtinsModule)) {
+          case Some x when (boolean) x.get() ->
+            // Ensure that builtins doesn't try and have codegen run on it.
+            updateModule(
+                    builtinsModule,
+                    u -> u.compilationStage(CompilationStage.AFTER_CODEGEN)
+            );
+          default -> {
+            builtins.initializeBuiltinsIr(this, freshNameSupply, passes);
+            updateModule(
+                    builtinsModule,
+                    u -> u.hasCrossModuleLinks(true)
+            );
+          }
+        }
```

Comment:
```suggestion
        if (serializationManager.deserialize(builtinsModule).getOrElse(() -> true)) {
          // Ensure that builtins doesn't try and have codegen run on it.
          updateModule(
                  builtinsModule,
                  u -> u.compilationStage(CompilationStage.AFTER_CODEGEN)
          );
        } else {
          builtins.initializeBuiltinsIr(this, freshNameSupply, passes);
          updateModule(
                  builtinsModule,
                  u -> u.hasCrossModuleLinks(true)
          );
        }
```

Reply:
The `switch` & `case` matches [original code](https://github.com/enso-org/enso/pull/7666/files#diff-3c1feecee162a04218aa14d830adc7c703a619e5c1528b3e9933eece8fa41947L120) more. Plus your suggestion contains a lambda function and I'd rather avoid lambdas.

Reply:
Yeah, I know. Didn't like Scala equivalent either ;) Mostly because having a simple guard in `case Some x when (boolean) x.get() ->` feels unnecessary.

Reply:
OK, rewritten to if: [9df39e5](https://github.com/enso-org/enso/pull/7666/commits/9df39e501677a1239a47c9ad3d9f34ab820a284a)